### PR TITLE
Java: Add `LCS` command (with IDX option)

### PR DESCRIPTION
**Pull request by GumpacG**
https://redis.io/docs/latest/commands/lcs/

*Description of changes:*
Implementing `LCS` with the `IDX` option requires some glide core changes due to the complex objects returned.

Return object of `LCS IDX` without `WITHMATCHLEN` option:
Screenshot 2024-06-25 at 2 11 34 PM

Return object of `LCS IDX` with `WITHMATCHLEN` option:
Screenshot 2024-06-25 at 2 11 46 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 1]
```diff
@@ -339,9 +340,65 @@ public interface StringBaseCommands {
...
+     *     </code> String contains a two-dimensional Long array that stores the pair of start and end
+     *     indices of the first and second Strings that match. An empty <code>Object</code> in the
+     *     <code>HashMap</code> is returned if the keys do not exist or have no common subsequences.
+     * @example
+     *     <pre>{@code
+     * // testKey1 = "abcd", testKey2 = "bcde"
+     * Map<String, Object> result = client.lcsIdx("testKey1", "testKey2").get();
+     * Map<String, Object> expectedLcsIdxObject =Map.of("matches", new Object[] {new Long[][] {{1L, 3L}, {0L, 2L}}},
+     *      "len", 3L);
+     * // result is equal to expectedLcsIdxObject
```
Comment (GumpacG) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:371]: I added a comment since user won't have access to `assertDeepEquals` and there is no great way to have a concise assert for `Map<String, Object>`

File: java/client/src/main/java/glide/api/models/commands/LcsOptions.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    public static final String WITHMATCHLEN_COMMAND_STRING = "WITHMATCHLEN";
+
+    /** Minimum length of matches to include in the result. */
+    private final Long minMatchLen;
+
+    /** Will include match lengths in the result if set to <code>true</code>. */
+    private boolean isWithMatchLen;
+
+    /** Sets <code>isWithMatchLen</code> to <code>true</code>. */
+    public static class LcsOptionsBuilder {
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/commands/LcsOptions.java:32]: Should we get rid of this line since we already document the function below?
```suggestion
    public static class LcsOptionsBuilder {
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 3]
```diff
@@ -5422,4 +5423,91 @@ public void lcs_with_len_option(BaseClient client) {
...
+
+        // LCS with only IDX and WITHMATCHLEN
+        expectedMatchesObject =
+                new Object[] {
+                    new Object[] {new Long[] {6L, 10L}, new Long[] {8L, 12L}, 5L},
+                    new Object[] {new Long[] {3L, 5L}, new Long[] {0L, 2L}, 3L}
+                };
+        result = client.lcsIdx(key1, key2, LcsOptions.builder().withMatchLen().build()).get();
+        assertDeepEquals(expectedMatchesObject, result.get("matches"));
+        assertEquals(8L, result.get("len"));
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:5477]: It looks like this is a duplicate block of the block above?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 4]
```diff
@@ -5422,4 +5423,91 @@ public void lcs_with_len_option(BaseClient client) {
...
+        // LCS with only IDX, and MINMATCHLEN
+        expectedMatchesObject =
+                new Object[] {
+                    new Long[][] {{6L, 10L}, {8L, 12L}},
+                };
+        result = client.lcsIdx(key1, key2, LcsOptions.builder().minMatchLen(4L).build()).get();
+        assertDeepEquals(expectedMatchesObject, result.get("matches"));
+        assertEquals(8L, result.get("len"));
+
+        // LCS with only IDX, MINMATCHLEN, and WITHMATCHLEN
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:5488]: nit
```suggestion
        // LCS with IDX, MINMATCHLEN, and WITHMATCHLEN
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 5]
```diff
@@ -4634,6 +4636,57 @@ public T lcsLen(@NonNull String key1, @NonNull String key2) {
...
     }
 
+    /**
+     * Returns the indices and length of the longest common subsequence between strings stored at
+     * <code>key1</code> and <code>
+     * key2</code>.
+     *
+     * @since Redis 7.0 and above.
+     * @apiNote When in cluster mode, <code>key1</code> and <code>key2</code> must map to the same
+     *     hash slot.
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:4646]: We don't need this warning for transactions
```suggestion
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 6]
```diff
@@ -4634,6 +4636,57 @@ public T lcsLen(@NonNull String key1, @NonNull String key2) {
...
+    }
+
+    /**
+     * Returns the indices and length of the longest common subsequence between strings stored at
+     * <code>key1</code> and <code>
+     * key2</code>.
+     *
+     * @since Redis 7.0 and above.
+     * @apiNote When in cluster mode, <code>key1</code> and <code>key2</code> must map to the same
+     *     hash slot.
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:4670]:
```suggestion
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 7]
```diff
@@ -5422,4 +5423,91 @@ public void lcs_with_len_option(BaseClient client) {
...
+        expectedMatchesObject =
+                new Object[] {new Object[] {new Long[] {6L, 10L}, new Long[] {8L, 12L}, 5L}};
+        result =
+                client
+                        .lcsIdx(key1, key2, LcsOptions.builder().minMatchLen(4L).withMatchLen().build())
+                        .get();
+        assertDeepEquals(expectedMatchesObject, result.get("matches"));
+        assertEquals(8L, result.get("len"));
+
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:5497]: Should we add test(s) that pass negative and/or 0 for the minMatchLen value?

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 8]
```diff
@@ -348,25 +348,25 @@ public interface StringBaseCommands {
...
      *     hash slot.
      * @see <a href="https://valkey.io/commands/lcs/">valkey.io</a> for details.
      * @param key1 The key that stores the first string.
      * @param key2 The key that stores the second string.
-     * @return A <code>HashMap<String, Object</code> containing the indices of longest common
-     *     subsequence between the 2 strings. The <code>Object</code> mapped to the <code>"matches"
-     *     </code> String contains a two-dimensional Long array that stores the pair of start and end
-     *     indices of the first and second Strings that match. An empty <code>Object</code> in the
-     *     <code>HashMap</code> is returned if the keys do not exist or have no common subsequences.
+     * @return A <code>Map&lt;String, Object&gt;</code> containing the indices of the longest common
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:359]: Just wanted to double check that using `&lt;` instead of `<` was intentional? I'm assuming that's how you need to represent `<` inside a code block?

**Review by aaron-congo — APPROVED** [review]

File: glide-core/src/client/value_conversion.rs [hunk 9]
```diff
@@ -859,6 +859,11 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
                 value_type: &Some(ExpectedReturnType::ArrayOfPairs),
             }),
         }),
+        b"LCS" => cmd.position(b"IDX")
+                   .map(|_| ExpectedReturnType::Map {
+            key_type: &Some(ExpectedReturnType::BulkString),
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:863]:
```suggestion
            key_type: &Some(ExpectedReturnType::SimpleString),
```
Those are hardcoded strings "matches" and "len", not binary strings.
Update tests then and lint rust code

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:359]: You can use just a `Map`

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 10]
```diff
@@ -339,9 +340,65 @@ public interface StringBaseCommands {
...
+     *     subsequence between the 2 strings and the length of the longest common subsequence. The
+     *     <code>Object</code> mapped to the <code>"matches"</code> map key contains a two-dimensional
+     *     <code>Long</code> array that stores the pair of start and end indices of the first and
+     *     second strings that match. An empty <code>Object</code> in the <code>Map</code> is returned
+     *     if the keys do not exist or have no common subsequences.
+     * @example
+     *     <pre>{@code
+     * // testKey1 = "abcd", testKey2 = "bcde"
+     * Map<String, Object> result = client.lcsIdx("testKey1", "testKey2").get();
+     * Map<String, Object> expectedLcsIdxObject = Map.of("matches", new Object[] {new Long[][] {{1L, 3L}, {0L, 2L}}},
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:369]: Why not?
```suggestion
     * Map<String, Object> expectedLcsIdxObject = Map.of("matches", new Long[][][] {{{1L, 3L}, {0L, 2L}}},
```
Is it 2D or 3D array?

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 8 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:359]: Yes, that how we can show `<` in html.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 11]
```diff
@@ -339,9 +340,65 @@ public interface StringBaseCommands {
...
+     *
+     * @since Redis 7.0 and above.
+     * @apiNote When in cluster mode, <code>key1</code> and <code>key2</code> must map to the same
+     *     hash slot.
+     * @see <a href="https://valkey.io/commands/lcs/">valkey.io</a> for details.
+     * @param key1 The key that stores the first string.
+     * @param key2 The key that stores the second string.
+     * @return A <code>Map&lt;String, Object&gt;</code> containing the indices of the longest common
+     *     subsequence between the 2 strings and the length of the longest common subsequence. The
+     *     <code>Object</code> mapped to the <code>"matches"</code> map key contains a two-dimensional
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:361]: I'd say this in an other words:
> The resulting map contains two keys: "matches" and "len":
> * "len" is mapped to the length of the common subsequence stored as `Long`.
> * "matches" mapped to a 3D array of .....

You need to describe what are these magic numbers in "matches". Maybe copy description from another client or source if you can find something. As an option you can refer to the example.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:371]: You can use `printf` and do a pretty print (not just numbers, but with text) of the values, so it will make the result self-described.
See my example in `functionList` and `functionStats`

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 10 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:369]: Please confirm that it is `Long[]`, but not `Object[]` - so you convert arrays

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 12]
```diff
@@ -339,9 +340,65 @@ public interface StringBaseCommands {
...
+     * <code>key1</code> and <code>key2</code>.
+     *
+     * @since Redis 7.0 and above.
+     * @apiNote When in cluster mode, <code>key1</code> and <code>key2</code> must map to the same
+     *     hash slot.
+     * @see <a href="https://valkey.io/commands/lcs/">valkey.io</a> for details.
+     * @param key1 The key that stores the first string.
+     * @param key2 The key that stores the second string.
+     * @param lcsOptions The {@link LcsOptions}.
+     * @return A <code>Map&lt;String, Object&gt;</code> containing the indices of the longest common
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:387]: same here

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 13]
```diff
@@ -4634,6 +4636,51 @@ public T lcsLen(@NonNull String key1, @NonNull String key2) {
...
 
+    /**
+     * Returns the indices and length of the longest common subsequence between strings stored at
+     * <code>key1</code> and <code>key2</code>.
+     *
+     * @since Redis 7.0 and above.
+     * @see <a href="https://valkey.io/commands/lcs/">valkey.io</a> for details.
+     * @param key1 The key that stores the first string.
+     * @param key2 The key that stores the second string.
+     * @return Command Response - A <code>Map&lt;String, Object&gt;</code> containing the indices of
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:4647]: And then copy updated description there

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 14]
```diff
@@ -5422,4 +5423,91 @@ public void lcs_with_len_option(BaseClient client) {
...
+        // setting string values
+        client.set(key1, "abcdefghijk");
+        client.set(key2, "defjkjuighijk");
+
+        // LCS with only IDX
+        Object expectedMatchesObject =
+                new Object[] {
+                    new Long[][] {{6L, 10L}, {8L, 12L}},
+                    new Long[][] {{3L, 5L}, {0L, 2L}}
+                };
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:5454]: Why not `Long[][][]`? Probably conversion is missing

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 15]
```diff
@@ -1891,6 +1893,20 @@ public CompletableFuture<Long> lcsLen(@NonNull String key1, @NonNull String key2
         return commandManager.submitNewCommand(LCS, arguments, this::handleLongResponse);
     }
 
+    @Override
+    public CompletableFuture<Map<String, Object>> lcsIdx(@NonNull String key1, @NonNull String key2) {
+        String[] arguments = new String[] {key1, key2, IDX_COMMAND_STRING};
+        return commandManager.submitNewCommand(LCS, arguments, this::handleMapResponse);
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:1899]: If you need to convert, probably do something like `handleFunctionListResponse`

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 16]
```diff
@@ -208,6 +209,16 @@ private static Object[] stringCommands(BaseTransaction<?> transaction) {
         String stringKey7 = "{StringKey}-7-" + UUID.randomUUID();
         String stringKey8 = "{StringKey}-8-" + UUID.randomUUID();
 
+        Map<String, Object> expectedLcsIdxObject =
+                Map.of("matches", new Object[] {new Long[][] {{1L, 3L}, {0L, 2L}}}, "len", 3L);
+
+        Map<String, Object> expectedLcsIdxWithMatchLenObject =
+                Map.of(
+                        "matches",
+                        new Object[] {new Object[] {new Long[] {1L, 3L}, new Long[] {0L, 2L}, 3L}},
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:218]: Here you have `new Object[] {new Object[] {new Long[] ...`, should be `Long[][][]` again

File: glide-core/src/client/value_conversion.rs (hunk 9 above)
Comment (Yury-Fridlyand) [reply glide-core/src/client/value_conversion.rs:863]: I think yes. We still need that in scope of binary string support.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 17]
```diff
@@ -339,9 +340,188 @@ public interface StringBaseCommands {
...
+     *
+     * @since Redis 7.0 and above.
+     * @apiNote When in cluster mode, <code>key1</code> and <code>key2</code> must map to the same
+     *     hash slot.
+     * @see <a href="https://valkey.io/commands/lcs/">valkey.io</a> for details.
+     * @param key1 The key that stores the first string.
+     * @param key2 The key that stores the second string.
+     * @return A <code>Map</code> containing the indices of the longest common subsequence between the
+     *     2 strings and the length of the longest common subsequence. The resulting map contains two
+     *     keys: "matches" and "len":
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:361]: There are 2 colons into one phrase. Maybe use a comma?
```suggestion
     *     keys, "matches" and "len":
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 18]
```diff
@@ -339,9 +340,188 @@ public interface StringBaseCommands {
...
+     *     <ul>
+     *       <li>"len" is mapped to the length of the longest common subsequence between the 2 strings
+     *           stored as <code>Long</code>.
+     *       <li>"matches" is mapped to a three dimensional <code>Long</code> array that stores pairs
+     *           of indices that represent the location of the common subsequences in the strings held
+     *           by <code>key1</code> and <code>key2</code>. For example, the sample result <code>
+     *           new Long[][][] {{{1L, 3L}, {0L, 2L}}}</code> indicates that the substring in <code>
+     *           key1</code> at index 1 to 3 matches the substring in <code>key2</code> at index 0 to
+     *           2.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:370]: > ... and it is the only common substring.

Maybe worth describing how location of the second LCS is represented in the response or show example with two, but I don't wont overload the doc.
Up to you - what do you feel is better.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 19]
```diff
@@ -339,9 +340,188 @@ public interface StringBaseCommands {
...
+     *     2 strings and the length of the longest common subsequence. The resulting map contains two
+     *     keys: "matches" and "len":
+     *     <ul>
+     *       <li>"len" is mapped to the length of the longest common subsequence between the 2 strings
+     *           stored as <code>Long</code>.
+     *       <li>"matches" is mapped to a three dimensional <code>Long</code> array that stores pairs
+     *           of indices that represent the location of the common subsequences in the strings held
+     *           by <code>key1</code> and <code>key2</code>. For example, the sample result <code>
+     *           new Long[][][] {{{1L, 3L}, {0L, 2L}}}</code> indicates that the substring in <code>
+     *           key1</code> at index 1 to 3 matches the substring in <code>key2</code> at index 0 to
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:369]: Wrap all numbers in docs with `code`

**Comment by Yury-Fridlyand** [comment]
Fix rust CI please

**Review by Yury-Fridlyand — APPROVED** [review]
It is OK to switch to `GlideString` for arguments now or after the merge.
Keep in mind that you need to update all tests and examples. (note: transaction does not support `GlideString` yet)

File: glide-core/src/client/value_conversion.rs [hunk 20]
```diff
@@ -859,6 +863,10 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
                 value_type: &Some(ExpectedReturnType::ArrayOfPairs),
             }),
         }),
+        b"LCS" => cmd.position(b"IDX").map(|_| ExpectedReturnType::Map {
+            key_type: &Some(ExpectedReturnType::SimpleString),
```
[the remark this answers is not in the document]
Comment (GumpacG) [reply glide-core/src/client/value_conversion.rs:867]: Comment from Yury: https://github.com/Bit-Quill/glide-for-redis/pull/386#discussion_r1649409472

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 21]
```diff
@@ -4634,6 +4637,205 @@ public T lcsLen(@NonNull String key1, @NonNull String key2) {
...
+     *     resulting map contains two keys, "matches" and "len":
+     *     <ul>
+     *       <li>"len" is mapped to the length of the longest common subsequence between the 2 strings
+     *           stored as <code>Long</code>.
+     *       <li>"matches" is mapped to a three dimensional <code>Long</code> array that stores pairs
+     *           of indices that represent the location of the common subsequences in the strings held
+     *           by <code>key1</code> and <code>key2</code>. For example,
+     *     </ul>
+     *
+     * @example If <code>key1</code> holds the string <code>"abcd123"</code> and <code>key2</code>
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:4756]: Now copy the improved TM examples to the command interface.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 22]
```diff
@@ -4634,6 +4637,205 @@ public T lcsLen(@NonNull String key1, @NonNull String key2) {
...
+     *          new Long[] {4L, 6L},
+     *          new Long[] {5L, 7L},
+     *          3L},
+     *      new Object[] {
+     *          new Long[] {1L, 3L},
+     *          new Long[] {0L, 2L},
+     *          3L}
+     *      }
+     * }</pre>
+     *     The result indicates that the first substring match is <code>"123"</code> in <code>key1
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:4770]: I guess the first match is "bcd". Can you try making them different lenght to distinguish? For example, key1 = "abcd123", key2 = "bcdef12"

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 21 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:4756]: I think we can omit adding it to the command interface because the example code will do a better job of explaining the result and it would be redundant to have two examples in the interface docs. I didn't add sample code to transaction because it is trickier to show the results.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 22 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: Addressed in [810b3df](https://github.com/Bit-Quill/glide-for-redis/pull/386/commits/810b3df1ae41b4acfac6522e41687f58f4900775). Is that what you meant?

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: Yes, thanks!

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 21 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: I mean, don't add second example, but replace existing - it is way more detailed.

**Review by acarbonetto — APPROVED** [review]

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 23]
```diff
@@ -446,6 +447,15 @@ protected Map<String, Map<String, Object>> handleFunctionStatsResponse(
         return response;
     }
 
+    /** Process a <code>LCS key1 key2 IDX</code> response */
+    protected Map<String, Object> handleLcsIdxResponse(Map<String, Object> response)
+            throws RedisException {
+        Long[][][] convertedMatchesObject =
+                cast3DArray((Object[]) (response.get("matches")), Long.class);
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/BaseClient.java:454]: NPE will be raised

[the remark this answers is not in the document]
Comment (GumpacG) [reply java/client/src/main/java/glide/api/BaseClient.java]: Unfortunately it returns null silently as is and does not throw an exception. I've added throwing an NPE for this case in [71301ee](https://github.com/Bit-Quill/glide-for-redis/pull/386/commits/71301ee3d96ce5c7ce3851f7f10f01c4fcd7f0e4)
